modules/currency-l0/src/test/scala/io/constellationnetwork/currency/l0/snapshot/SnapshotEventsPublisherDaemonSuite.scala

```diff
@@ -0,0 +1,123 @@
+      implicit0(sp: SecurityProvider[IO]) <- SecurityProvider.forAsync[IO]
+      h = Hasher.forJson[IO]
+      kp <- KeyPairGenerator.makeKeyPair[IO].toResource
+    } yield (s, h, sp, kp)
+
+  private def waitUntilQueueSize(queue: Queue[IO, _], expectedSize: Int, attempt: Int = 0): IO[Unit] =
+    for {
+      size <- queue.size
+      res <-
+        if (size == expectedSize) {
+          IO.unit
+        } else if (attempt >= 30) {
+          IO.raiseError(new TimeoutException)
+        } else {
+          IO.sleep(1.second).flatMap(_ => waitUntilQueueSize(queue, expectedSize, attempt + 1))
+        }
+    } yield res
+
+  test("should publish events to gossip without duplicates") { res =>
+    implicit val (s, h, sp, keyPair) = res
+
+    implicit val m = NoOpMetrics.make
+
+    implicit val e = {
+      implicit val noopEncoder: Encoder[DataUpdate] = new Encoder[DataUpdate] {
+        final def apply(a: DataUpdate): Json = Json.Null
+      }
+      CurrencySnapshotEvent.encoder
+    }
+
+    val blockEvent1 = signedBlockGen.sample.map(BlockEvent(_)).get
+    val blockEvent2 = signedBlockGen.sample.map(BlockEvent(_)).get
+
+    val selfId = PeerId(Hex("0000000000000000"))
+
+    val consensusConfig = ConsensusConfig(
+      timeTriggerInterval = 10.seconds,
+      declarationTimeout = 10.seconds,
+      declarationRangeLimit = NonNegLong(10),
+      lockDuration = 10.seconds,
+      eventCutter = EventCutterConfig(
+        maxBinarySizeBytes = PosInt(5000000)
+      )
+    )
+
+    for {
+      consensusStorage <- ConsensusStorage.make[
+        IO,
+        CurrencySnapshotEvent,
+        CurrencySnapshotKey,
+        CurrencySnapshotArtifact,
+        CurrencySnapshotContext,
+        CurrencySnapshotStatus,
+        CurrencyConsensusOutcome,
+        CurrencyConsensusKind
+      ](consensusConfig)
+
+      _ <- consensusStorage.addEvents(Map(selfId -> List((Ordinal(Generation(PosLong(1)), Counter(PosLong(1))), blockEvent1))))
+
+      rumorQueue: Queue[IO, Hashed[RumorRaw]] <- Queue.unbounded[IO, Hashed[RumorRaw]]
+
+      l1OutputQueue <- Queue.unbounded[IO, Option[CurrencySnapshotEvent]]
+      _ <- l1OutputQueue.offer(Some(blockEvent1))
+      _ <- l1OutputQueue.offer(Some(blockEvent1))
+      _ <- l1OutputQueue.offer(Some(blockEvent2))
+      _ <- l1OutputQueue.offer(None)
+
+      events = Stream.fromQueueNoneTerminated(l1OutputQueue)
+      gossip <- Gossip.make[IO](rumorQueue, selfId, Generation(PosLong(1)), keyPair)
+      _ <- SnapshotEventsPublisherDaemon.make(gossip, events, consensusStorage).spawn.start
+      _ <- waitUntilQueueSize(l1OutputQueue, 0)
+      _ <- waitUntilQueueSize(rumorQueue, 1)
+      rumorQueueSize <- rumorQueue.size
+      maybeRumor <- rumorQueue.tryTake
+      maybeEitherBlockEvent = maybeRumor.map { rumor =>
+        val cur = rumor.signed.value.content.hcursor
+        cur.downField("value").get[BlockEvent]("BlockEvent")
+      }
```

Comment:
what is the Rumor type? do you have to manually play with json to extract the type of the rumor?